src/util/integerfactoring.h

```diff
@@ -0,0 +1,241 @@
+/* Include Guards */
+#ifndef INTEGERFACTORING_H
+#define INTEGERFACTORING_H
+
+
+/* Includes */
+#include <stdio.h>
+#include <stdint.h>
+#include <stdint-gcc.h>
+
+
+/* Defines */
+
+
+
+/* C++ Extern "C" Guard */
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+
+/* Data Structure Prototypes & Typedefs */
+struct GA_FACTOR_LIST;
+typedef struct GA_FACTOR_LIST GA_FACTOR_LIST;
+
+
+
+/* Data Structures */
+
+/**
+ * @brief The GA_FACTOR_LIST struct.
+ * 
+ * Contains the list of distinct prime factors of a 64-bit unsigned integer, as
+ * well as the powers of those factors.
+ * 
+ * There can be at most 15 such distinct factors, since the product of the
+ * first 16 primes (2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53) exceeds
+ * the maximum unsigned number of 2^64-1. Moreover, there can be at most 63
+ * factors all together, since 2^64 exceeds 2^64-1, so only an 8-bit number is
+ * required to store the powers.
+ * 
+ * The 15th (last) element of the factor list is always 0 and has power 0,
+ * and serves as a sort of sentinel.
+ */
+
+struct GA_FACTOR_LIST{
+	uint64_t f[16];/* Factors */
+	uint8_t  p[16];/* Powers of factors */
+};
+
+
+
+/* Functions */
+
+/**
+ * @brief Count trailing zeros of a 64-bit integer.
+ * 
+ * @param [in] n  The integer whose trailing zero count is to be computed.
+ * @return     If n != 0, returns trailing zero count; Else returns 64.
+ */
+
+int      gaICtz(uint64_t n);
```

Comment:
The functions to count trailing/leading zeros and the modular arithmetic are pretty much only for factoring so they should probably be statics.  There may be others, but I'm not 100% sure.

The interface to the rest of the library (i.e. functions that will be called from other files) should stay in the header.